### PR TITLE
Replace hook keyring.backends to keyring.

### PR DESCRIPTION
**Pull request by ktong**
From keyring 13.0, uses entrypoints to read keyring.backends from metadata file entry_points.txt.

### CODE REVIEW

**Comment by arossert** [comment]
This is required for the latest keyring to work properly on Windows.

**Comment by earonesty** [comment]
Seems to work on our win 10, osx mojave/catalina + ubu/arch builds.

**Comment by vossen-adobe** [comment]
Can confirm this fix works on multiple windows versions using keyring-21.1.0 and pyinstaller 3.6.  Would be awesome to get this merged!

**Comment by lyneca** [comment]
Would be great to see this merged :)

**Comment by tlx494** [comment]
Keen to see this merged as well, hopefully happens soonish? Thanks :)

**Comment by oryandunn** [comment]
> Can confirm this fix works on multiple windows versions using keyring-21.1.0 and pyinstaller 3.6. Would be awesome to get this merged!

I've tried manually implementing this fix, but I've not had success.  I'm using Python 3.6. Keyring 21.2.1 and PyInstaller 3.6.  You can see the test code I am working with here:
https://github.com/jaraco/keyring/issues/439

After I remove the old hook file and added the new one as given in this PR in site-packages/PyInstaller/hooks folder, and rerun pyinstaller --onefile print_credentials.py, the app still has the error win32cred not defined.

I'm really keen to get this working, any assistance would be appreciated.

**Comment by Legorooj** [comment]
@oryandunn I don't know if this PR will get merged; I'd like to move the hook to keyring itself, now we support that. It that doesn't happen, I'll prioritise this PR.

**Comment by vossen-adobe** [comment]
@Legorooj  any update here? This fix works for us, but wondering if this has been changed as you mentioned?

**Review by bwoodsend — COMMENTED** [review]
This seems sensible to me. It'll drag in every available back-end but such is the nature of the *backend* model.

File: PyInstaller/hooks/hook-keyring.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from PyInstaller.utils.hooks import collect_submodules, copy_metadata
+
+hiddenimports = collect_submodules('keyring.backends')
+
+# keyring uses entrypoints to read keyring.backends from metadata file entry_points.txt.
+datas = copy_metadata('keyring')
```
Comment (bwoodsend) [inline PyInstaller/hooks/hook-keyring.py:15]: Surely for a backend to declare itself, its metadata needs to be collected too?

Comment (rokm) [reply PyInstaller/hooks/hook-keyring.py:15]: The backends are "registered" by having their entry points listed in the (main package's) `entry_points.txt`, i.e.:
```
[keyring.backends]
KWallet = keyring.backends.kwallet
SecretService = keyring.backends.SecretService
Windows = keyring.backends.Windows
chainer = keyring.backends.chainer
macOS = keyring.backends.OS_X
```
That's why `keyring`'s metadata needs to be collected.

This PR is essentially the same as #5245 (I didn't notice it before), except that it also "upgrades" the hook from `keyring.backends` to `keyring` (which does not seem to be strictly necessary, though).

Comment (bwoodsend) [reply PyInstaller/hooks/hook-keyring.py:15]: Ahh, I always assumed entry points were organised the other way around. 

> it also "upgrades" the hook from keyring.backends to keyring

Unless it's possible to write code that uses `keyring` but doesn't use `keyring.backends` (even indirectly) then this has no effect really.

**Comment by bwoodsend** [comment]
Well I see no reason not to merge this. If `keyring` choose to provide their own hooks in the future then theirs have priority over ours so having it here too does no harm. @ktong Are you still with us? Could you rebase your PR to fix the merge conflict, `flake8`-ify the hook to appease our miserable linter?

**Comment by Legorooj** [comment]
Closing in favour of #5245.

**Pull request closed without merging**